src/future/batch_semaphore.rs

```diff
@@ -57,17 +226,26 @@ struct BatchSemaphoreState {
     //
     // (4) closed ==> waiters.is_empty()
     waiters: VecDeque<Arc<Waiter>>,
-    permits_available: usize,
+    permits_available: PermitsAvailable,
+    // TODO: should there be a clock for the close event?
```

Comment:
Yeah, think we need to add a causal dependency from the close event to subsequent `Closed` error returns.  But we can do that in a followup.

Reply:
Yes. I tried to implement this, but ran into issues with getting a clock in the `close` method, because it may happen as part of the cleanup process/drop handlers. We talked with @sarsko a bit about re-organising how drops happen, which would unblock this (adding a clock for closing).

Reply:
Should we track this in an issue?